certora/specs/Gates.spec

```diff
@@ -0,0 +1,146 @@
+    function _.canReceiveShares(address user) external => ghostStatusCanReceiveShares[user] expect bool;
+    function _.canSendAssets(address user) external =>  ghostStatusCanSendAssets[user] expect bool;
+    function _.canReceiveAssets(address user) external => ghostStatusCanReceiveAssets[user] expect bool;
+
+    function _.supply(MorphoMarketV1Adapter.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external
+        => summaryMorphoMarketV1Supply(marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MorphoMarketV1Adapter.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external
+        => summaryMorphoMarketV1Withdraw(marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+
+    function _.deposit(uint256 assets, address receiver) external =>
+        summaryMorphoVaultV1Deposit(assets, receiver) expect uint256 ALL;
+    function _.withdraw(uint256 assets, address receiver, address owner) external =>
+        summaryMorphoVaultV1Withdraw(assets, receiver, owner) expect uint256 ALL;
+
+    function _.allocate(bytes data, uint256 assets, bytes4, address) external => DISPATCHER(true);
+    function _.deallocate(bytes data, uint256 assets, bytes4, address) external => DISPATCHER(true);
+    function _.realAssets() external => DISPATCHER(true);
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+}
+
+function summaryMorphoMarketV1Supply(MorphoMarketV1Adapter.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256) {
+    assert shares == 0;
+    assert data.length == 0;
+    uint256 returnedShares;
+    require (ghostStatusCanReceiveAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can receive assets");
+    ERC20Helper.safeTransferFrom(marketParams.loanToken, onBehalf, MorphoMarketV1Adapter.morpho, assets);
+    return (returnedShares, shares);
+}
+
+function summaryMorphoMarketV1Withdraw(MorphoMarketV1Adapter.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) returns (uint256, uint256) {
+    assert shares == 0;
+    uint256 returnedShares;
+    require (ghostStatusCanSendAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can send assets");
+    ERC20Helper.safeTransferFrom(marketParams.loanToken, MorphoMarketV1Adapter.morpho, onBehalf, assets);
+    return (assets, returnedShares);
+}
+
+function summaryMorphoVaultV1Deposit(uint256 assets, address receiver) returns uint256 {
+    uint256 shares;
+    require (ghostStatusCanReceiveAssets[MorphoMarketV1Adapter.morpho], "require that MorphoMarketV1 can receive assets");
+    ERC20Helper.safeTransferFrom(currentContract.asset, MorphoVaultV1Adapter, MorphoMarketV1Adapter.morpho, assets);
```

Comment:
Wondering if we should not summarize it as 2 transfers: one from the adapter to the vault v1 and one from the vault v1 to market v1. This way we would catch the fact that vault v1 also sees its balance change